tests/importer/importer.cpp

```diff
@@ -0,0 +1,761 @@
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, replaceModel)
+{
+    // This test shows how a model instance can be replaced in the import library by the user,
+    // and will be used to resolve imports.
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("importer/generic_no_source.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    auto wrongSourceModel = libcellml::Model::create("wrong");
+
+    auto rightSourceModel = libcellml::Model::create("source");
+    rightSourceModel->addComponent(libcellml::Component::create("a"));
+    rightSourceModel->addComponent(libcellml::Component::create("b"));
+    rightSourceModel->addComponent(libcellml::Component::create("c"));
+    rightSourceModel->addComponent(libcellml::Component::create("d"));
+
+    // Add a model manually to the library, including the URL that it will replace in future imports.
+    EXPECT_TRUE(importer->addModel(wrongSourceModel, "i_dont_exist.cellml"));
+    // Can't add to the same URL key more than once.
+    EXPECT_FALSE(importer->addModel(rightSourceModel, "i_dont_exist.cellml"));
+
+    // Use replaceModel instead.
+    EXPECT_FALSE(importer->replaceModel(rightSourceModel, "not_in_library"));
+    EXPECT_TRUE(importer->replaceModel(rightSourceModel, "i_dont_exist.cellml"));
+
+    importer->resolveImports(model, "");
+
+    EXPECT_EQ(size_t(0), importer->issueCount());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, getNonexistentModel)
+{
+    auto importer = libcellml::Importer::create();
+    auto model = libcellml::Model::create("model");
+    importer->addModel(model, "howdy");
+    EXPECT_EQ(nullptr, importer->library("bonjour"));
+}
+
+TEST(Importer, getListOfDependencies)
+{
+    // This test shows how a model can be interrogated to return a list of its dependencies/URLs.
+    // These then become the URL "keys" that must be supplied to the importer, if their actual location
+    // should not be used or is inaccessible.
+
+    // Note that in order to resolve the dependencies further than the first generation, the models will
+    // need to be instantiated and saved anyway.  The external dependencies are those which, during the
+    // resolution for this model, were sourced from locations outside the library.  They will have since
+    // been saved to the importer library, and subsequent calls to resolveImports for those same imported
+    // files will not change the dependencies list.
+
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("importer/diamond.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    EXPECT_EQ(size_t(3), importer->externalDependencyCount());
+    for (size_t e = 0; e < importer->externalDependencyCount(); ++e) {
+        auto info = importer->externalDependency(e);
```

Comment:
I understand that this PR is really about moving code around, but `Importer::externalDependency()` currently returns a `std::pair<std::string, std::string>` object, which is not neat. Ideally, we would return a pointer to a class that has only two methods to retun each of the strings in the original pair. (This is what is done in PR #631 with the `AnalyserModel` class for example.)

See issue #681.

Reply:
To be addressed later as an issue (if still needed).